grafana_client/client.py

```diff
@@ -119,72 +126,145 @@ def construct_api_url():
         def __request_runner(url, json=None, data=None, headers=None, accept_empty_json=False):
-            __url = "%s%s" % (self.url, url)
+            __url = self._make_url(url)
             # Sanity checks.
-            if json is not None and not isinstance(json, (dict, list)):
-                raise TypeError(
-                    f"JSON request payload has invalid shape. "
-                    f"Accepted are dictionaries and lists. "
-                    f"The type is: {type(json)}"
+            GrafanaClient._ensure_valid_json_arg(json)
+
+            try:
+                r = self.s.request(
+                    item.lower(),
+                    __url,
+                    json=json,
+                    data=data,
+                    headers=headers,
+                    auth=self.auth,
+                    verify=self.verify,
+                    timeout=self.timeout,
                 )
-            runner = getattr(self.s, item.lower())
-            r = runner(
-                __url,
-                json=json,
-                data=data,
-                headers=headers,
-                auth=self.auth,
-                verify=self.verify,
-                timeout=self.timeout,
-            )
-            if r.status_code >= 400:
-                try:
-                    response = r.json()
-                except ValueError:
-                    response = r.text
-                message = response["message"] if isinstance(response, dict) and "message" in response else r.text
-
-                if 500 <= r.status_code < 600:
-                    raise GrafanaServerError(
-                        r.status_code,
-                        response,
-                        "Server Error {0}: {1}".format(r.status_code, message),
-                    )
-                elif r.status_code == 400:
-                    raise GrafanaBadInputError(response)
-                elif r.status_code == 401:
-                    raise GrafanaUnauthorizedError(response)
-                elif 400 <= r.status_code < 500:
-                    raise GrafanaClientError(
-                        r.status_code,
-                        response,
-                        "Client Error {0}: {1}".format(r.status_code, message),
-                    )
-
-            # `204 No Content` responses have an empty response body,
-            # so it doesn't decode well from JSON.
-            if r.status_code == 204:
-                return None
-
-            # The "Tempo" data source responds with text/plain.
-            content_type = r.headers.get("Content-Type", "")
-            if content_type.startswith("text/"):
-                return r.text
+            except Timeout as e:
+                raise GrafanaTimeoutError(0, None, str(e)) from e
+            except HTTPError as e:
+                # this branch is to make sure we don't leak anything bellow Grafana Client! like ProtocolError.
+                raise GrafanaException(0, None, str(e)) from e
```

Comment:
Excellent, well done! Just a suggestion about wording on the inline comment.
```suggestion
                # Make sure to not leak any exception types of the requests implementation.
                raise GrafanaException(0, None, str(e)) from e
```